File: addTwoNumbers.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */

var addTwoNumbers = function (l1, l2) {
  // reverse the two linked lists
  const num1 = reverseLinkedList(l1);
  const num2 = reverseLinkedList(l2);

  // sum the numbers & reverse order
  const sum = computeReverseSum(num1, num2);

  // fill result array w/ linked list nodes
  let result = [];
  sum.forEach((digit) => {
    result.push(new ListNode(digit));
  });

  // link nodes
  result.forEach((node, index) => {
    node.next = result[index + 1];
  });

  // return LL head
  return result[0];
};

function reverseLinkedList(ll) {
  // create an empty "stack"
  const stack = [];
  // treat result as string for easier reversing
  let result = '';

  let cur = ll;
  stack.push(cur);
  // build up stack
  while (cur.next) {
    cur = cur.next;
    stack.push(cur);
  }

  // pop items off stack and add to result, reversing the order
  while (stack.length) {
    const node = stack.pop();
    result += node.val;
  }

  // cast result back to number and return
  return Number(result);
}

function computeReverseSum(num1, num2) {
  let sum = num1 + num2;
  // cast to string, reverse order of digits, then cast back to numbers
  let result = String(sum)
    .split('')
    .reverse()
    .map((char) => Number(char));

  return result;
}

File: add_two_numbers.py
# You are given two non-empty linked lists representing two non-negative integers. The digits are stored in reverse order and each of their nodes contain a single digit. Add the two numbers and return it as a linked list.

# You may assume the two numbers do not contain any leading zero, except the number 0 itself.

# Example:

# Input: (2 -> 4 -> 3) + (5 -> 6 -> 4)
# Output: 7 -> 0 -> 8
# Explanation: 342 + 465 = 807.






# Definition for singly-linked list.
class ListNode:
  def __init__(self, val=0, next=None):
    self.val = val
    self.next = next




def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
  carry_over = 0
  node1 = l1
  node2 = l2
  
  no_next = False
  list_start = None
  previous = None
  
  while no_next == False:
    num = node1.val + node2.val
    num += carry_over
    carry_over = num // 10
    num %= 10
    
    if previous != None:
      previous.next = ListNode(num, None)
      previous = previous.next
        
    
    if list_start == None:
      list_start = ListNode(num, None)
      previous = list_start
      
    
    if node1.next == None and node2.next == None and carry_over == 0:
      no_next = True
        
        
    if node1.next != None:
      node1 = node1.next
    else:
      node1 = ListNode(0, None)
        
        
    if node2.next != None:
      node2 = node2.next
    else:
      node2 = ListNode(0, None)
          
  return list_start